S3: Basic unstructured file support

flash1293 opened the pull request:
## What

Fixes https://github.com/airbytehq/airbyte/issues/31208

<img width="566" alt="Screenshot 2023-10-10 at 13 23 03" src="https://github.com/airbytehq/airbyte/assets/1508364/90e77dfb-15e8-4ea4-95bb-8b967a3f0e43">

## Testing

To test, I spun up a local minio server using `docker run -p 9000:9000 -p 9001:9001 minio/minio server /data --console-address ":9001"`

To set up syncable data:
* Navigate to the admin console (http://localhost:9001/login) `minioadmin:minioadmin` for access
* Create a bucket
* Uploade some files 
* Create an access key with the following policy:
```
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "s3:*"
   ],
   "Resource": [
    "arn:aws:s3:::*"
   ]
  }
 ]
}
```
* Configure the S3 source with http://host.docker.internal:9000 as the endpoint

## How

* Add a new option to the file format configuration for unstructured formats
* Add a new parser to implement the new file format
* In the parser, delegate to the unstructured library `partition` method to extract information from the file
* Turn the returned array of elements into markdown (if the file is markdown already, just use it directly)
* Return a record with content and file id (just the file name in this case but could be different for other sources)
* If a file has the wrong format, it is treated as a parsing error

Limitations:
* For incremental syncs, Airbyte can only track up to 10k files - after this, it will apply a heuristic and only load files newer than a configurable number of dates
* For append+dedup syncs, stale chunks won't be removed if a file gets smaller than it was before (this is described in the documentation)

## Problems

* Check reads and parses a file currently which can be very slow if these files are large (e.g. 6MB PDF)
  * I will improve this on a follow-up PR as it needs changes to the CDK => https://github.com/airbytehq/airbyte/pull/31281

## Open questions
* Which file types to support? Technically, more file types are "for free" as the unstructured lib is doing the heavy lifting
  * IMHO, starting with markdown, pdf and docx is fine for now
* The markdown rendering is very basic currently, but the element structure of unstructured doesn't give us much to work with (for example the level of a heading). Is this good enough?
  * IMHO yes
* How large can we realistically go with chunk sizes? Need to check with platform team.
  * The biggest limitation is the maximum column size in the destination. 4M seems to be a safe choice for now, we can play with this later
* Single file size limit: The setup can't be streamed for a single file so it needs to fit into memory in its entirety. We don't grant a lot of memory to source connectors, easiest "handling" is to just let it crash trying to load a file that's too big to fit. Not sure whether this is good enough, we could add a check and skip files that are too large. Is this necessary?
  * IMHO just crashing is OK for now
* OCR - should we use an OCR strategy to allow extracting text from images in PDFs? Allow images as input format in general? The downside is the increased processing costs
* Markdown conversion loses some available meta information. Depending on the downstream processing, this might not be the best possible approach. We could add an output format option to the file format (markdown or json) or add two columns for both formats and rely on platform column selection to remove the once that's not needed. What's the right approach here?
  * IMHO we can start with just the markdown and defer that to later
* Do we want to support doc (separately from docx)? It requires libreoffice which I don't know whether we want to add as a dependency
  * IMHO we can defer that to later
* Should we do chunking or truncate a record if it gets too large? This would limit the usability, but make things a lot simpler (I can also imagine it's annoying in the destination to work with chunks all the time)
  * IMHO we can start with doing chunking - it's experimental so it should be OK to break it later on

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/2695eceee96cab15e76b1bd2428e615b5d1ffb38/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`2695eceee9`](https://github.com/airbytehq/airbyte/pull/31209/commits/2695eceee96cab15e76b1bd2428e615b5d1ffb38)) - ❌

⏲️  Total pipeline duration: 19mn54s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ❌        |
| Code format checks                                        | ❌        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697029478/2695eceee96cab15e76b1bd2428e615b5d1ffb38/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:2695eceee96cab15e76b1bd2428e615b5d1ffb38)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

flash1293 commented:
The acceptance test errors also happen on master (not caused by this PR): https://connectors.airbyte.com/files/generated_reports/test_summary/source-s3/index.html

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/a61a8c1be79ac70947791ae2de2faa05e9157f18/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`a61a8c1be7`](https://github.com/airbytehq/airbyte/pull/31209/commits/a61a8c1be79ac70947791ae2de2faa05e9157f18)) - ❌

⏲️  Total pipeline duration: 18mn59s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ❌        |
| Code format checks                                        | ✅        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697034037/a61a8c1be79ac70947791ae2de2faa05e9157f18/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:a61a8c1be79ac70947791ae2de2faa05e9157f18)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

aaronsteers commented:
> ## Problems
> * Check reads 10 files currently which can be very slow if these files are large (e.g. 6MB PDF)
>   
>   * I will improve this on a follow-up PR as it needs changes to the CDK => [File CDK: Allow configuration of parsed records during check and discover from parser #31281](https://github.com/airbytehq/airbyte/pull/31281)

Noted. And I think agree this is probably acceptable for a first pass, as discussed.

> ## Open questions
> * Which file types to support? Technically, more file types are "for free" as the unstructured lib is doing the heavy lifting
>   
>   * IMHO, starting with markdown, pdf and docx is fine for now

Sounds very reasonable. Less surface area to support means we can better focus on edge cases (if/when those arise) for a smaller number of possible file types. I'm also fine with excluding PDF in the first version if it makes sense to exclude, since it's the only one of these that requires OCR (image recognition) and _probably_ is an order of magnitude slower than the other purely text-based processing methods.

> * The markdown rendering is very basic currently, but the element structure of unstructured doesn't give us much to work with (for example the level of a heading). Is this good enough?
>   
>   * IMHO yes

Yes - and... I just checked in on my request here:

- https://github.com/Unstructured-IO/unstructured/issues/1354#issuecomment-1733720212

Was apparently closed as resolved here:

- https://github.com/Unstructured-IO/unstructured/pull/1268

@flash1293 - Could this help at all with giving us richer markdown info? (In this pass or in a future one.) I haven't closely looked at their implementation or spec as of yet.

> * How large can we realistically go with chunk sizes? Need to check with platform team.
>   
>   * The biggest limitation is the maximum column size in the destination. 4M seems to be a safe choice for now, we can play with this later

👍 

> * Single file size limit: The setup can't be streamed for a single file so it needs to fit into memory in its entirety. We don't grant a lot of memory to source connectors, easiest "handling" is to just let it crash trying to load a file that's too big to fit. Not sure whether this is good enough, we could add a check and skip files that are too large. Is this necessary?
>   
>   * IMHO just crashing is OK for now

👍 

> * OCR - should we use an OCR strategy to allow extracting text from images in PDFs? Allow images as input format in general? The downside is the increased processing costs

👍 - Long-term, I imagine we'd expose some of these nobs/dials to the users when they are using a parser that requires OCR. Perhaps instead of exposing all internals, we could reduce to "max quality", "max speed", or "balanced".

> * Markdown conversion loses some available meta information. Depending on the downstream processing, this might not be the best possible approach. We could add an output format option to the file format (markdown or json) or add two columns for both formats and rely on platform column selection to remove the once that's not needed. What's the right approach here?

Do you mean PDF->Markdown and Word->Markdown? 

In the first pass, I'm inclined to keep simple and just output the markdown text. But this is dependent on the specific nature of the issues/shortcomings...

>   * IMHO we can start with just the markdown and defer that to later

👍 

> * Do we want to support doc (separately from docx)? It requires libreoffice which I don't know whether we want to add as a dependency
>   
>   * IMHO we can defer that to later

Totally okay to keep doc (the older, pre-xml format) for later addition.

> * Should we do chunking or truncate a record if it gets too large? This would limit the usability, but make things a lot simpler (I can also imagine it's annoying in the destination to work with chunks all the time)
>   
>   * IMHO we can start with doing chunking - it's experimental so it should be OK to break it later on

I'm okay with chunking or truncating at this phase.

Argument for chunking:

- Most likely to deliver all the text data, regardless of text size.

Argument for truncating:

- Doesn't require us to have a robust chunking algorithm yet.

Argument for doing neither (yet):

- Shows us where the systems will break - specifically the platform and destinations.
- Doesn't add more complexity (yet) in terms of spec and/or record schema.
- Sets us up nicely to let to be deferred to destination-side or platform-side handling - which are both possible future outcomes.

flash1293 commented:
> Could this help at all with giving us richer markdown info? (In this pass or in a future one.) I haven't closely looked at their implementation or spec as of yet

Good callout, I think I missed this. Maybe the information is available already and I just don't access it properly. I will look into that.

flash1293 commented:
> Argument for doing neither (yet):

I like your reasoning here (plus the fact that it will work fine for a lot of cases). Wdyt about simplifying this PR and not doing the chunking?

clnoll commented on airbyte-integrations/connectors/source-s3/source_s3/v4/unstructured_parser.py:
```diff
@@ -0,0 +1,94 @@
+            markdown = self.read_file(fp, logger)
+            if not markdown:
+                return []
+            chunks = [markdown[i : i + self.MAX_SIZE_PER_CHUNK] for i in range(0, len(markdown), self.MAX_SIZE_PER_CHUNK)]
+            yield from [
+                {
+                    "content": chunk,
+                    "chunk_number": i,
+                    "no_of_chunks": len(chunks),
+                    "id": f"{file.uri}_{i}",
+                }
+                for i, chunk in enumerate(chunks)
+            ]
```
Feels like this should yield chunks instead of yielding from the list.

```suggestion
            chunk_number = 0
            no_of_chunks = ceil(len(markdown) / self.MAX_SIZE_PER_CHUNK)
            for chunk in chunked(markdown, self.MAX_SIZE_PER_CHUNK):
                yield  {
                    "content": chunk,
                    "chunk_number": chunk_number,
                    "no_of_chunks": no_of_chunks,
                    "id": f"{file.uri}_{i}",
                }
                chunk_number += 1
```

aaronsteers replied:
Good catch, @clnoll. The list comprehension causes all items to be collected before any are yielded.

Little-known trick, there's also a less-used [generator comprehension](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Generators_and_Comprehensions.html#Creating-your-own-generator:-generator-comprehensions) - with almost identical syntax.

```suggestion
            chunks = (markdown[i : i + self.MAX_SIZE_PER_CHUNK] for i in range(0, len(markdown), self.MAX_SIZE_PER_CHUNK))
            yield from (
                {
                    "content": chunk,
                    "chunk_number": i,
                    "no_of_chunks": len(chunks),
                    "id": f"{file.uri}_{i}",
                }
                for i, chunk in enumerate(chunks)
            )
```

Same inline syntax as the original, but should perform similarly to @clnoll's non-blocking proposal, which emits records chunks one at a time.

flash1293 replied:
As we decided to go without chunks for now, this got removed - still thanks for bringing it up, I didn't know about the generator comprehensions

clnoll commented on airbyte-integrations/connectors/source-s3/source_s3/v4/unstructured_parser.py:
```diff
@@ -0,0 +1,94 @@
+        with stream_reader.open_file(file, self.file_read_mode, None, logger) as fp:
+            markdown = self.read_file(fp, logger)
+            if not markdown:
+                return []
```
To keep the return type of this consistent, this should return an iterator instead of an iterable, e.g. `yield from []`.

flash1293 replied:
Got removed when removing chunks

clnoll commented on airbyte-integrations/connectors/source-s3/source_s3/v4/unstructured_parser.py:
```diff
@@ -0,0 +1,94 @@
+        elements = partition(file=file, metadata_filename=file_name)
+        return self._render_markdown(elements)
+
+    def _render_markdown(self, elements: List[Any]) -> str:
```
I see that we need to keep 2 copies of the file in memory since `unstructured`'s `partition` function [returns a list](https://github.com/Unstructured-IO/unstructured/blob/8ab40c20c143ca3f809c830f3c0214a35939fbe4/unstructured/partition/pdf.py#L448), but when we call `self._render_markdown` it feels like we'll actually temporarily have up to 3 copies. I think we can do better than that by modifying `self._render_markdown` -

```suggestion
    def _render_markdown(self, elements: List[Any]) -> str:
        for el in elements:
            yield self._convert_to_markdown(el)
            yield "\n\n"
```
We'd then have to collect all the strings together in `parse_records` and strip the final `"\n\n"`, which isn't the prettiest thing in the world, but if this seems worth doing maybe we can find a cleaner way.

I'm also wondering, since both PDF and DOCX documents have a concept of pages, whether there are any libraries that would be able to break on the page boundary. In that case we could parse one page at a time, which may even avoid downloading the entire file at once depending on its implementation. Obviously that could interfere with the quality of `unstructured`'s parsing, but it may be something to think about if we're feeling like there are memory issues.

flash1293 replied:
I replaced this with a generator comprehension which shouldn't have the problem.

> I'm also wondering, since both PDF and DOCX documents have a concept of pages, whether there are any libraries that would be able to break on the page boundary

That's an interesting idea, I'm not sure whether it's possible or not to stream through a PDF or DOCX without keeping the full thing in memory. I'm also not sure how common this use case is (multi GB pdfs seem possible, but rare). IMHO we should keep it simple for the first version and rely on the lib that's making things simpler for us and break out of this abstraction if we get user feedback that it's an important feature.

clnoll reviewed (commented):
Nice @flash1293! A few questions/comments.

aaronsteers commented on docs/integrations/sources/s3.md:
```diff
@@ -237,11 +237,29 @@ The Avro parser uses the [Fastavro library](https://fastavro.readthedocs.io/en/l
+
+One record will be emitted for each document. If the content of the document is longer than 4MB, the document will be split up into multiple records, each with a maximum size of 4MB. The fields "chunk_number" (starting at 0) and "no_of_chunks" will be added to each record to indicate the order of the chunks and the total number of chunks. The field "id" will be added to each record containing the ...
+
+If a file with another format is encountered, it is skipped. The sync will not be stopped in this case.
```
I think I agree that if a file is unreadable, it should probably stop the sync. This then delegates the exclusion of non-qualifying types to a glob pattern - or else the selection of a file type also more tightly constrains the glob pattern to just qualifying file extensions.

For example:

- If we have format set to PDF and there's a CSV incorrectly suffixed with `.pdf`, this will fail. Because our file-type-aware glob pattern (`/path/to/myfolder/**/*.pdf`) picked up a file that was unreadable.
- If we have format set to PDF and there's a CSV in the same directory, the CSV will be skipped. Because our file-type-aware glob pattern (`/path/to/myfolder/**/*.pdf`) was not concerned with files suffixed by `.csv`.

I'm using "glob pattern" as one concept, but how we implement may actually be a combination of other processes and patterns, for loops, etc.

flash1293 commented on airbyte-integrations/connectors/source-s3/unit_tests/v4/test_unstructured_parser.py:
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from unittest.mock import MagicMock, mock_open, patch
```
As this parser is part of the s3 source for now, I would like to defer this until I move it over into CDK proper, wdyt?

clnoll replied:
Ah right - that's fine.

flash1293 commented:
Thanks for your great comments @aaronsteers and @clnoll - I adjusted the PR and changed the following things:
* Removed the chunking
* Fail on encountering file with wrong type
* I'm not sure about the "id" field - currently it's identical with the `_ab_source_file_url` field, but there are two reasons why I think it makes sense to have it:
  * It's not an internal metadata field, it is explicitly meant to be used as a primary key
  * It holds the file url right now and for S3, but we might pick something else for other sources

Let me know what you think!

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/a27c1152e4c956fb4336e3a1875950d2e7ca799c/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`a27c1152e4`](https://github.com/airbytehq/airbyte/pull/31209/commits/a27c1152e4c956fb4336e3a1875950d2e7ca799c)) - ✅

⏲️  Total pipeline duration: 16mn13s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ✅        |
| Code format checks                                        | ✅        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697116352/a27c1152e4c956fb4336e3a1875950d2e7ca799c/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:a27c1152e4c956fb4336e3a1875950d2e7ca799c)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/9d66e3e4b352aaf400f886e9a534ee06d89782fc/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`9d66e3e4b3`](https://github.com/airbytehq/airbyte/pull/31209/commits/9d66e3e4b352aaf400f886e9a534ee06d89782fc)) - ✅

⏲️  Total pipeline duration: 17mn33s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ✅        |
| Code format checks                                        | ✅        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697186671/9d66e3e4b352aaf400f886e9a534ee06d89782fc/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:9d66e3e4b352aaf400f886e9a534ee06d89782fc)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/c1cc4d122d7be66473e2fefaa2f89ed6e21e368f/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`c1cc4d122d`](https://github.com/airbytehq/airbyte/pull/31209/commits/c1cc4d122d7be66473e2fefaa2f89ed6e21e368f)) - ❌

⏲️  Total pipeline duration: 17mn51s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ✅        |
| Code format checks                                        | ❌        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697192688/c1cc4d122d7be66473e2fefaa2f89ed6e21e368f/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:c1cc4d122d7be66473e2fefaa2f89ed6e21e368f)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/7f2ded80e72ba57fb489673897f597b0fd4bc48e/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`7f2ded80e7`](https://github.com/airbytehq/airbyte/pull/31209/commits/7f2ded80e72ba57fb489673897f597b0fd4bc48e)) - ❌

⏲️  Total pipeline duration: 17mn14s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ✅        |
| Code format checks                                        | ❌        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697194148/7f2ded80e72ba57fb489673897f597b0fd4bc48e/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:7f2ded80e72ba57fb489673897f597b0fd4bc48e)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

aaronsteers reviewed (approved):
I completed a full code readthrough. Besides what we previously discussed, I think the only other thing that pops out to me is the question of output schema, such as column names, incremental key, and primary key. I put a comment below that we may want to try to match the file_uri column name rather than introducing the new "id" column. 

Similarly, we could consider a column name like "document_markdown" or "file_markdown" or "markdown_contents" - but I also do think "contents" works. I didn't see schema for last-modified timestamps, but that may already be internally handled by the CDK.



Approving with these comments as non-blocking. This is a great first iteration. 🙏

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/6608dc53242444c019e5c525229531b383e42c69/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`6608dc5324`](https://github.com/airbytehq/airbyte/pull/31209/commits/6608dc53242444c019e5c525229531b383e42c69)) - ❌

⏲️  Total pipeline duration: 05mn22s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697445149/6608dc53242444c019e5c525229531b383e42c69/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:6608dc53242444c019e5c525229531b383e42c69)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/6608dc53242444c019e5c525229531b383e42c69/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`6608dc5324`](https://github.com/airbytehq/airbyte/pull/31209/commits/6608dc53242444c019e5c525229531b383e42c69)) - ❌

⏲️  Total pipeline duration: 06mn36s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697450983/6608dc53242444c019e5c525229531b383e42c69/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:6608dc53242444c019e5c525229531b383e42c69)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/6608dc53242444c019e5c525229531b383e42c69/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`6608dc5324`](https://github.com/airbytehq/airbyte/pull/31209/commits/6608dc53242444c019e5c525229531b383e42c69)) - ❌

⏲️  Total pipeline duration: 06mn44s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697452258/6608dc53242444c019e5c525229531b383e42c69/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:6608dc53242444c019e5c525229531b383e42c69)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/d14d6a8b1c9c4805b905f771646923a29d0d7ced/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`d14d6a8b1c`](https://github.com/airbytehq/airbyte/pull/31209/commits/d14d6a8b1c9c4805b905f771646923a29d0d7ced)) - ❌

⏲️  Total pipeline duration: 04mn25s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697460679/d14d6a8b1c9c4805b905f771646923a29d0d7ced/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:d14d6a8b1c9c4805b905f771646923a29d0d7ced)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

airbyte-oss-build-runner commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/495f428322d74450f02cfea094909552886716cc/airbyte-integrations/connectors/source-s3/icon.svg" width="40" height="40"> source-s3 test report (commit [`495f428322`](https://github.com/airbytehq/airbyte/pull/31209/commits/495f428322d74450f02cfea094909552886716cc)) - ✅

⏲️  Total pipeline duration: 16mn29s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅       |
| Unit tests                                                | ✅       |
| Acceptance tests                                          | ✅       |
| Code format checks                                        | ✅       |
| Validate metadata for source-s3                           | ✅       |
| Connector version semver check                            | ✅       |
| Connector version increment check                         | ✅       |
| QA checks                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_markdown-parser/1697528679/495f428322d74450f02cfea094909552886716cc/source-s3/4.0.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:495f428322d74450f02cfea094909552886716cc)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```
